sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -238,60 +240,49 @@ public void populateDisplayData(DisplayData.Builder builder) {
   }
 
   /**
-   * Writes all the elements in a bundle using a {@link FileBasedWriter} produced by the
-   * {@link FileBasedSink.FileBasedWriteOperation} associated with the {@link FileBasedSink}.
+   * Writes all the elements in a bundle using a {@link Writer} produced by the
+   * {@link WriteOperation} associated with the {@link FileBasedSink} with windowed writes enabled.
    */
-  private class WriteBundles extends DoFn<T, FileResult> {
-    // Writer that will write the records in this bundle. Lazily
-    // initialized in processElement.
-    private FileBasedWriter<T> writer = null;
-    private BoundedWindow window = null;
+  private class WriteWindowedBundles extends DoFn<T, FileResult> {
+    private Map<KV<BoundedWindow, PaneInfo>, Writer<T>> windowedWriters;
```

Comment:
In particular while this will work for our built-in windows, users can write their own window functions, and we do not insist that they provide hashCode - all we insist is that they provide a Coder. Hence the reason I used Coder.structuredValue to encode the windows.